core/corehttp/ipns_hostname.go

```diff
@@ -19,8 +19,8 @@ func IPNSHostnameOption() ServeOption {
 			defer cancel()
 
 			host := strings.SplitN(r.Host, ":", 2)[0]
-			if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-				r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+			if p, err := n.Namesys.Resolve(ctx, host, 0); err == nil {
+				r.URL.Path = p.String() + r.URL.Path
```

Comment:
I was going to say that for `(c)` we do not need to add a header. We already set 

```
X-IPFS-Path: /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/index.html
```

So what the gateway can do is -- after redirecting the first time, setting `X-IPFS-Path` -- it could check for `X-IPFS-Path` and ignore the url's path if so. i.e. stop after (2) below.

Gateway resolve:
1. set `path = /`
2. check X-IPFS-Path, if there is one, set `path = <x-ipfs-path>`. DONE!
3. see hostname -- resolve dnslink. set `path = <dnslink-path>`
4. see URL (relative) path. set `path = path + url.path` DONE!

(not sure if the above is a fully correct characterization in light of @Luzifer's IPNS solution, which we should do).

This -- the `X-IPFS-Path` approach -- is more flexible, i think, as some services may be constrained in their "URL prettiness rules" and the header would be a ground truth of what to resolve.